src/rules/checkTagNames.js

```diff
@@ -54,12 +110,103 @@ export default iterateJsdoc(({
       });
   }
 
-  for (const jsdocTag of jsdoc.tags) {
+  const isInAmbientContext = (subNode) => {
+    return subNode.type === 'Program' ?
+      context.getFilename().endsWith('.d.ts') :
+      Boolean(subNode.declare) || isInAmbientContext(subNode.parent);
+  };
+
+  const tagIsRedundantWhenTyped = (jsdocTag) => {
+    if (!typedTagsUnnecessaryOutsideDeclare.has(jsdocTag.tag)) {
+      return false;
+    }
+
+    if (jsdocTag.description.trim() === 'default' || jsdocTag.name === 'default') {
+      return false;
+    }
+
+    if (context.getFilename().endsWith('.d.ts') && node.parent.type === 'Program') {
+      return false;
+    }
+
+    if (isInAmbientContext(node)) {
+      return false;
+    }
+
+    return true;
+  };
+
+  const reportWithTypeRemovalFixer = (message, jsdocTag, tagIndex, additionalTagChanges) => {
+    utils.reportJSDoc(message, jsdocTag, () => {
+      if (jsdocTag.description.trim()) {
+        utils.changeTag(jsdocTag, {
+          postType: '',
+          type: '',
+          ...additionalTagChanges,
+        });
+      } else {
+        utils.removeTag(tagIndex, {
+          removeEmptyBlock: true,
+        });
+      }
+    }, true);
+  };
+
+  const tagRedundantlyChecksParamOrReturnTyped = (jsdocTag, tagIndex) => {
+    if (!typedTagsParamReturn.has(jsdocTag.tag)) {
+      return false;
+    }
+
+    if (!jsdocTag.description.trim()) {
+      reportWithTypeRemovalFixer(`'@${jsdocTag.tag}' without a description is redundant when using a type system.`, jsdocTag, tagIndex);
+      return true;
+    }
+
+    if (jsdocTag.type) {
+      reportWithTypeRemovalFixer(`Describing the type of '@${jsdocTag.tag}' is redundant when using a type system.`, jsdocTag, tagIndex);
+      return true;
+    }
+
+    return false;
+  };
+
+  const checkTagForTypedValidity = (jsdocTag, tagIndex) => {
+    if (typedTagsAlwaysUnnecessary.has(jsdocTag.tag)) {
+      reportWithTypeRemovalFixer(`'@${jsdocTag.tag}' is redundant when using a type system.`, jsdocTag, tagIndex, {
+        postTag: '',
+        tag: '',
+      });
+      return true;
+    }
+
+    if (tagIsRedundantWhenTyped(jsdocTag)) {
+      reportWithTypeRemovalFixer(`'@${jsdocTag.tag}' is generally redundant outside of \`declare\` contexts when using a type system.`, jsdocTag, tagIndex);
```

Comment:
Is it generally redundant or always redundant? If *generally*, when is it not redundant?

Reply:
Yeah I struggled with this phrasing 😕. In some situations, such as "ambient" (type declaration) contexts (the `"declare"` comment), they are allowed. https://github.com/gajus/eslint-plugin-jsdoc/pull/1007#discussion_r1154616106

Reply:
The phrasing already mentions being redundant outside of `declare` contexts, so if that is the only case, I'd think we can drop "generally". If it is not redundant in other contexts, I'd hope we could either catch those contexts or if necessary not remove those particular tags.

Reply:
Hmm yeah good point - other than the conversation in https://github.com/gajus/eslint-plugin-jsdoc/pull/1007#discussion_r1154487772, these are all ambient/declaration contexts... I'll remove that for now since I don't see any non-DefinitelyTyped usage for that logic.